File: rd-text/src/main/kotlin/com/jetbrains/rider/rdtext/RdTextChangeEx.kt
package com.jetbrains.rider.rdtext

import com.jetbrains.rider.rdtext.intrinsics.RdTextChange
import com.jetbrains.rider.rdtext.intrinsics.RdTextChangeKind

fun RdTextChange.delta(): Int {
    if (RdTextChangeKind.Reset == kind) throw UnsupportedOperationException()
    return new.length - old.length
}

fun RdTextChange.reverse(): RdTextChange {
    val newKind = when (kind) {
        RdTextChangeKind.Insert -> RdTextChangeKind.Remove
        RdTextChangeKind.Remove -> RdTextChangeKind.Insert
        RdTextChangeKind.Replace -> RdTextChangeKind.Replace
        RdTextChangeKind.Reset -> throw UnsupportedOperationException()
    }

    return RdTextChange(newKind, startOffset, new, old, fullTextLength - delta())
}

fun RdTextChange.assertDocumentLength(current: Int) {
    if (RdTextChangeKind.Reset == kind) return

    val actualLength = current + delta()
    val expectedLength = fullTextLength
    if (actualLength != expectedLength)
        throw IllegalStateException("Expected document length: $expectedLength, but actual: $actualLength")
}

File: rd-net/RdFramework/Text/RdTextChangeEx.cs
using System;
using JetBrains.Rd.Text.Intrinsics;

namespace JetBrains.Rd.Text
{
  public static class RdTextChangeEx
  {
    public static int GetDelta(this RdTextChange that) => that.New.Length - that.Old.Length;

    public static void AssertDocumentLength(this RdTextChange that, int currentLen)
    {
      if (that.Kind != RdTextChangeKind.Reset && that.FullTextLength != -1)
      {
        var actual = currentLen + that.GetDelta();
        var expected = that.FullTextLength;
        if (actual != expected)
          throw new InvalidOperationException($"Expected the document size: {expected}, but actual: {actual}.");
      }
    }
  }
}